src/main/java/com/enonic/xp/app/welcome/WelcomePageScriptBean.java

```diff
@@ -230,18 +325,110 @@ public String getStatisticsApiUrl()
+    {
+        final ApplicationInstallResultJson result = new ApplicationInstallResultJson();
+
+        try
+        {
+            ContextBuilder.from( ContextAccessor.current() ).authInfo(
+                AuthenticationInfo.create().principals( RoleKeys.ADMIN ).user( SUPER_USER ).build() ).build().runWith( () -> {
+
+                final Application application = this.applicationServiceSupplier.get().installGlobalApplication( url, sha512 );
+
+                result.setApplication( toApplicationJson( application ) );
+
+            } );
+
+        }
+        catch ( Exception e )
+        {
+            final String failure = "Failed to process application from " + url;
+            LOG.error( failure, e );
+
+            result.setFailure( failure );
+        }
+        return result;
+    }
+
+    private ApplicationJson toApplicationJson( final Application application )
+    {
+        ApplicationKey applicationKey = application.getKey();
+        ApplicationJson.Builder builder =
+            ApplicationJson.create().application( application ).description( getApplicationDescription( applicationKey ) ).iconAsBase64(
+                getApplicationIconAsBase64( applicationKey ) );
+
+        Resource resource = resourceServiceSupplier.get().getResource( ResourceKey.from( applicationKey, "/webapp/webapp.js" ) );
+        if ( resource.exists() )
+        {
+            String webappUrl = ServletRequestUrlHelper.createUri( ServletRequestHolder.getRequest(), "/webapp/" + application.getKey() );
+            builder.webappUrl( webappUrl );
+        }
+
+        List<String> adminToolsUris = getApplicationAdminToolsUris( applicationKey );
+        adminToolsUris.forEach( builder::addAdminToolsUrl );
+
+        return builder.build();
+    }
+
+    private <V> V lock( Object key, Callable<V> callable )
+    {
+        final Lock lock = LOCK_STRIPED.get( key );
+        try
+        {
+            if ( lock.tryLock( 30, TimeUnit.MINUTES ) )
+            {
+                try
+                {
+                    return callable.call();
+                }
+                catch ( Exception e )
+                {
+                    throw Exceptions.unchecked( e );
+                }
+                finally
+                {
+                    lock.unlock();
+                }
+            }
+            else
+            {
+                throw new RuntimeException( "Failed to acquire application service lock for application [" + key + "]" );
+            }
+        }
+        catch ( InterruptedException e )
+        {
+            throw new RuntimeException( "Failed to acquire application service lock for application [" + key + "]", e );
+        }
+    }
+
+    private List<TemplateApplicationJson> mustParseJsonFile( String filePath )
+    {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try
+        {
+            return objectMapper.readValue( new FileInputStream( filePath ), new TypeReference<>()
```

Comment:
FileInputStream must be opened in try block

Reply:
Like the one on line 407 ? )

Reply:
@rymsha ?

Reply:
I gave an example how to do it https://github.com/enonic/app-xp-welcome/pull/263#discussion_r1357874079